### DISCUSSION TITLE
How to not serialize some elements when calling graph.toJSON()

### DISCUSSION THREAD

**Discussion by TomPast** [clientIO/joint#2072] [Q&A] [answered]
### Introduction

Hi there,

I have three types of elements in my app, lets say: ElementA, ElementB and Link. But when I call `graph.toJSON()`, I just want to have my ElementA and Link to be in the JSON.

I know I could do that by filtering my `graph.toJSON()` result, but is there a way to do that directly in the `ElementB`  `toJSON()` definition like that:

```Javascript
export const ElementB = dia.Element.define(
  'app.ElementB',
  {
    size: {
      width: 250,
      height: 300,
    },
    attrs: {
        ...attrs
    },
  },
  {
    markup: [
      ...markup
    ],
    // Should not be serialized in JointJS JSON.
    toJSON() {
      return undefined; // Is there a value I could put here to not insert this element in the `graph.toJSON()`
    },
  },
);
```

Lets say I have an `ElementA`and an `ElementB` in my graph, the JSON graph will look like : 
```JSON
{
    "cells": [
        {
            "z": 1,
            "id": "d7dde79f-0144-405f-b17d-dsdsddsdsds",
            "data": {
            },
            "size": {
                "width": 75,
                "height": 75
            },
            "type": "app.ElementA",
            "angle": 0,
            "attrs": {},
            "position": {
                "x": 220,
                "y": 150
            }
        },
        null <-- Here I would like to have no value because its an ElementB
    ]
}

```

I just tried different values like `undefined`, `null` or `""`. But then I can't load back my JSON graph with `graph.fromJSON()` without having this error `TypeError: Cannot read properties of null (reading 'type')`

### Does your question relate to JointJS or JointJS+. Select both if applicable.

JointJS

**Comment by kumilingus** [comment, marked as answer]
It's not supported. And I don't think it's a good idea to allow it. The `element` that should be omitted from JSON could be the target of a `link`. If the `link` were exported, it would make the JSON invalid.

Also, some of the plugins (e.g. `ui.Clipboard`) use `cell.toJSON()` internally to copy cells. That would be also broken by returning `null`.

Instead, try doing this:

*cherry pick what you want to export:*
```js
const cellsToExport = [elementA1, elementA2, linkA1A2];
const json =  { cells: cellsToExport.map(cell => cell.toJSON()); }
```

**Reply by TomPast** [reply]
Thanks for the reply ! I didn't know that `cell.toJSON()` was called by some plugins internally.

I did it like that (because I'm sure its not the source or the target of any links) :
```Javascript
const cellsToExport = graph.getCells().filter(cell => cell.attributes.type !== 'app.ElementB');
const json = { cells: cellsToExport.map(cell => cell.toJSON()) };
```
